hibernate-core/src/main/java/org/hibernate/query/sql/internal/NativeQueryImpl.java

```diff
@@ -644,6 +657,18 @@ public KeyedResultList<R> getKeyedResultList(KeyedPage<R> page) {
 	}
 
 	protected SelectQueryPlan<R> resolveSelectQueryPlan() {
+		if ( resultEntityClass != null ) {
+			boolean exists = false;
+			for ( ResultBuilder existing : resultSetMapping.getResultBuilders() ) {
+				if ( resultEntityClass == existing.getJavaType() ) {
+					exists = true;
+					break;
+				}
+			}
+			if ( !exists ) {
+				addEntity( (Class<R>) resultEntityClass, LockMode.READ );
+			}
+		}
```

Comment:
I was more thinking of creating a `ResultSetMapping` on demand instead of altering the existing one to avoid interference with user configured mappings.

```suggestion
		final ResultSetMapping mapping;
		if ( resultJavaType != null && resultSetMapping.isDynamic() && resultSetMapping.getNumberOfResultBuilders() == 0 ) {
			mapping = ResultSetMapping.resolveResultSetMapping( originalSqlString, true, getSessionFactory() );
			mapping.addResultBuilder(
				getSessionFactory().getMappingMetamodel().isEntityClass( resultJavaType )
					? Builders.entityCalculated( StringHelper.unqualify( entityType.getName() ), entityType.getName(), LockMode.READ, getSessionFactory() )
					: Builders.resultClassBuilder( resultClass, getSessionFactory() )
			);
		}
		else {
			mapping = resultSetMapping;
		}
```

Reply:
Amended as you suggested, please note `resultEntityClass` always is entity class or null.